In [1]:
!cat spambase.DOCUMENTATION

In [2]:
!cat spambase.names

In [3]:
import pandas as pd
import numpy as np
import sklearn
import re

In [4]:
spam_data = pd.read_csv("https://archive.ics.uci.edu/ml/machine-learning-databases/spambase/spambase.data", header=None)

Gotta get those headers, and I'll be damned if I'm going to copy and paste them all. I mean, I *could* train the filter on it without names, but how is that readable?

In [5]:
spam_data.head()

0     1     2   3     4     5     6     7     8     9  ...    48     49  \
0  0.00  0.64  0.64   0  0.32  0.00  0.00  0.00  0.00  0.00 ...  0.00  0.000   
1  0.21  0.28  0.50   0  0.14  0.28  0.21  0.07  0.00  0.94 ...  0.00  0.132   
2  0.06  0.00  0.71   0  1.23  0.19  0.19  0.12  0.64  0.25 ...  0.01  0.143   
3  0.00  0.00  0.00   0  0.63  0.00  0.31  0.63  0.31  0.63 ...  0.00  0.137   
4  0.00  0.00  0.00   0  0.63  0.00  0.31  0.63  0.31  0.63 ...  0.00  0.135   

   50     51     52     53     54   55    56  57  
0   0  0.778  0.000  0.000  3.756   61   278   1  
1   0  0.372  0.180  0.048  5.114  101  1028   1  
2   0  0.276  0.184  0.010  9.821  485  2259   1  
3   0  0.137  0.000  0.000  3.537   40   191   1  
4   0  0.135  0.000  0.000  3.537   40   191   1  

[5 rows x 58 columns]

In [6]:
names = open("spambase.names")
names.readline()

'| SPAM E-MAIL DATABASE ATTRIBUTES (in .names format)\n'

In [7]:
column_names = re.compile(r"(^\w+.):")
name_text = names.readlines()

In [8]:
columns = []
for line in name_text:
    if column_names.match(line):
        columns.append(column_names.match(line).groups(0)[0])

In [9]:
columns

['word_freq_make',
 'word_freq_address',
 'word_freq_all',
 'word_freq_3d',
 'word_freq_our',
 'word_freq_over',
 'word_freq_remove',
 'word_freq_internet',
 'word_freq_order',
 'word_freq_mail',
 'word_freq_receive',
 'word_freq_will',
 'word_freq_people',
 'word_freq_report',
 'word_freq_addresses',
 'word_freq_free',
 'word_freq_business',
 'word_freq_email',
 'word_freq_you',
 'word_freq_credit',
 'word_freq_your',
 'word_freq_font',
 'word_freq_000',
 'word_freq_money',
 'word_freq_hp',
 'word_freq_hpl',
 'word_freq_george',
 'word_freq_650',
 'word_freq_lab',
 'word_freq_labs',
 'word_freq_telnet',
 'word_freq_857',
 'word_freq_data',
 'word_freq_415',
 'word_freq_85',
 'word_freq_technology',
 'word_freq_1999',
 'word_freq_parts',
 'word_freq_pm',
 'word_freq_direct',
 'word_freq_cs',
 'word_freq_meeting',
 'word_freq_original',
 'word_freq_project',
 'word_freq_re',
 'word_freq_edu',
 'word_freq_table',
 'word_freq_conference',
 'char_freq_;',
 'char_freq_(',
 'char_freq_[',
 '

Oh, yeah. We're cooking with oil now.

With gas? What's that saying? Huh...

In [10]:
len(columns)

57

In [11]:
spam_data[57]

0       1
1       1
2       1
3       1
4       1
5       1
6       1
7       1
8       1
9       1
10      1
11      1
12      1
13      1
14      1
15      1
16      1
17      1
18      1
19      1
20      1
21      1
22      1
23      1
24      1
25      1
26      1
27      1
28      1
29      1
       ..
4571    0
4572    0
4573    0
4574    0
4575    0
4576    0
4577    0
4578    0
4579    0
4580    0
4581    0
4582    0
4583    0
4584    0
4585    0
4586    0
4587    0
4588    0
4589    0
4590    0
4591    0
4592    0
4593    0
4594    0
4595    0
4596    0
4597    0
4598    0
4599    0
4600    0
Name: 57, dtype: int64

In [12]:
columns.append("is_spam")
spam_data.columns = columns
spam_data.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64             0   
1            0.21               0.28           0.50             0   
2            0.06               0.00           0.71             0   
3            0.00               0.00           0.00             0   
4            0.00               0.00           0.00             0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail   ...     char_freq_;  char_freq_(  \
0             0.00            0.00   ...            0.00        0.000   
1             0.00            0.94   ...            0.00        0.132   
2             0.64            0.25   ...            0.01        0.143   
3             0.31            0.63   ...            0.00        0.137   
4             0.31            0.63   ...            0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0            0        0.778        0.000        0.000   
1            0        0.372        0.180        0.048   
2            0        0.276        0.184        0.010   
3            0        0.137        0.000        0.000   
4            0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  is_spam  
0                       278        1  
1                      1028        1  
2                      2259        1  
3                       191        1  
4                       191        1  

[5 rows x 58 columns]

In [13]:
spam_data_np = spam_data.values

In [14]:
from sklearn.cross_validation import train_test_split

In [15]:
training, testing = train_test_split(spam_data_np, test_size=0.4)

In [16]:
training.shape

(2760, 58)

In [17]:
training_dep = training[:,:-1]

In [18]:
training_dep.shape

(2760, 57)

In [19]:
training_target = training[:, -1:]

In [20]:
training_target.shape

(2760, 1)

In [21]:
from sklearn.naive_bayes import GaussianNB

In [22]:
bayes_filter = GaussianNB()
bayes_filter.fit(training_dep, training_target)

/Users/gordon/.pyenv/versions/sandbox/lib/python3.4/site-packages/sklearn/naive_bayes.py:150: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


GaussianNB()

In [23]:
predictions = bayes_filter.predict(testing[:,:-1])

In [24]:
testing_dep = testing[:, :-1]
testing_target = testing[:, -1:]

In [25]:
from sklearn import metrics
from sklearn.cross_validation import cross_val_score
def test_display_prediction(correct, predictions):
    print(metrics.classification_report(correct, predictions))
    print(metrics.confusion_matrix(correct, predictions))
    print(metrics.f1_score(correct, predictions))
test_display_prediction(testing_target, predictions)

             precision    recall  f1-score   support

        0.0       0.95      0.71      0.82      1125
        1.0       0.68      0.94      0.79       716

avg / total       0.84      0.80      0.80      1841

[[804 321]
 [ 43 673]]
0.787134502924


In [26]:
from sklearn.linear_model import Perceptron
perceptron_classifier = Perceptron()

In [27]:
perceptron_classifier.fit(training_dep, training_target)

/Users/gordon/.pyenv/versions/sandbox/lib/python3.4/site-packages/sklearn/linear_model/stochastic_gradient.py:321: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


Perceptron(alpha=0.0001, class_weight=None, eta0=1.0, fit_intercept=True,
      n_iter=5, n_jobs=1, penalty=None, random_state=0, shuffle=False,
      verbose=0, warm_start=False)

In [28]:
perceptron_predict = perceptron_classifier.predict(testing_dep)

In [29]:
test_display_prediction(testing_target, perceptron_predict)

             precision    recall  f1-score   support

        0.0       0.61      1.00      0.76      1125
        1.0       0.00      0.00      0.00       716

avg / total       0.37      0.61      0.46      1841

[[1125    0]
 [ 716    0]]
0.0


/Users/gordon/.pyenv/versions/sandbox/lib/python3.4/site-packages/sklearn/metrics/metrics.py:1771: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)
/Users/gordon/.pyenv/versions/sandbox/lib/python3.4/site-packages/sklearn/metrics/metrics.py:1771: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no predicted samples.
  'precision', 'predicted', average, warn_for)


In [30]:
from sklearn.linear_model import LogisticRegression

So, actually, we want to use a logit regression. I very dimly recall this from my days in an econometrics class. Logistic regressions are particularly well-adapted to cases where we're trying to predict a binary outcome.

In [31]:
logit_classifier = LogisticRegression()

In [32]:
logit_classifier.fit(training_dep, training_target)

/Users/gordon/.pyenv/versions/sandbox/lib/python3.4/site-packages/sklearn/preprocessing/label.py:125: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, penalty='l2', random_state=None, tol=0.0001)

In [33]:
logit_predictions = logit_classifier.predict(testing_dep)
logit_predictions

array([ 0.,  0.,  0., ...,  0.,  0.,  0.])

In [34]:
test_display_prediction(testing_target, logit_predictions)

             precision    recall  f1-score   support

        0.0       0.93      0.94      0.93      1125
        1.0       0.90      0.89      0.90       716

avg / total       0.92      0.92      0.92      1841

[[1053   72]
 [  77  639]]
0.895585143658


Wow, that actually worked fairly well. Let's try to use a different Bayes classifier.

In [35]:
from sklearn.naive_bayes import MultinomialNB

In [36]:
multi_bayes = MultinomialNB()

In [37]:
multi_bayes.fit(training_dep, training_target)

/Users/gordon/.pyenv/versions/sandbox/lib/python3.4/site-packages/sklearn/naive_bayes.py:301: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True)

In [38]:
multi_bayes_predictions = multi_bayes.predict(testing_dep)
multi_bayes_predictions.shape

(1841,)

In [39]:
test_display_prediction(testing_target, multi_bayes_predictions)

             precision    recall  f1-score   support

        0.0       0.82      0.83      0.83      1125
        1.0       0.73      0.72      0.73       716

avg / total       0.79      0.79      0.79      1841

[[935 190]
 [200 516]]
0.725738396624


So, for a Gaussian Naive Bayes, we got a score of about 0.80, for a Logit Regression we got a score of about 0.90, and for a Multinomial Naive Bayes, we got a score of about 0.70. Good job, Logit Regression.